rfc/redis-removal.md

```diff
@@ -0,0 +1,171 @@
+| **SIG / WG**   | Platform SIG, Operations SIG |
+
+## Overview
+
+To help reduce the technical debt of the project and improve the operational
+overhead of running Spinnaker, Redis needs to be deprecated and removed from
+being a persistence option in favor of using SQL as the baseline storage
+technology. This RFC documents a process and timeline for doing so.
+
+### Goals and Non-Goals
+
+Goals:
+
+*   Provide an inventory of Spinnaker services that still need SQL support added
+*   Define a process that all services can use for migrating datastores without
+    downtime
+*   Provide a timeline for the final switch to SQL, along with criteria to
+    decide if the timeline should be delayed
+
+Non-Goals:
+
+*   Debating if SQL is the best replacement default persistence backend to
+    standardize on
+*   Documenting and solving all technical issues for each service to migrate;
+    those will be handled separately whereas this document concerns itself
+    primarily with the larger processes and timeline
+*   Identification of all new processes for maintaining services on a SQL
+    backend; these will be handled separately
+*   Prescription of what SQL vendor to use: Current functionality has been built
+    using MySQL; support for other vendors will be addressed separately
+
+## Case Against Redis
+
+Redis was introduced into the project as an easy way to get started. It's
+phenomenally flexible, and a single, common data store simplifies the
+operational story of the large operational footprint that Spinnaker has.
+Further, it was useful in that its flexibility meant services could be rapidly
+iterated on, as no schema was needed from the start.
+
+Unfortunately, Redis is difficult to operationalize as a persistent datastore,
+which many of the services treat Redis as. Community members are often surprised
+when Redis restarts to find Spinnaker has lost in-flight work. Operating Redis
+beyond a cache is not a skill that many organizations have. Worse yet, the level
+of effort to support a highly-available, persistent story for Redis is very high
+and littered with edge cases. For example, when tested at Netflix, both Redis
+Cluster and Sentinel regularly encountered incorrect master failover behavior
+that would lead to corrupted clusters or lost data.
+
+Furthermore, developing applications on Redis is typically a foreign concept and
+offers a confusing touch point for anyone looking to get involved in
+contributing.
+
+Having a singular datastore for Spinnaker is still necessary, as it eases the
+burden of onboarding. Now that Spinnaker is no longer a young project, the need
+for fast service iteration has largely dissapated. SQL was chosen as the
+replacement because modern SQL is flexible and well understood.
+
+## Service Inventory
+
+Most of the larger services already have support for SQL.
+
+* Full support: ✔️
+* Partial support: ~
+* No support: ✗
+
+| | |
+|-|-|
+| Clouddriver | ✔️ |
```

Comment:
Is it your belief that every cloud provider will work with the MySQL backend? That was not my impression, but I'd be happy to hear it...

Reply:
Ah, you're right. There are a couple things that need to be ironed out, but other cloud providers _can_ go onto SQL. A few people have managed to get k8s on SQL working.

Reply:
I know that there were some issues that had to be fixed to get Kubernetes to work, so it wouldn't surprise me if similar issues were seen with other cloud providers.